include/deal.II/optimization/line_minimization.h

```diff
@@ -0,0 +1,501 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2018 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE.md at
+//    the top level directory of deal.II.
+//
+//---------------------------------------------------------------
+
+#ifndef dealii_line_minimization_h
+#define dealii_line_minimization_h
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/numbers.h>
+#include <deal.II/base/utilities.h>
+
+#include <boost/optional.hpp>
+
+#include <errno.h>
+#include <sys/stat.h>
+
+#include <fstream>
+#include <string>
+
+DEAL_II_NAMESPACE_OPEN
+
+using namespace dealii;
+
+/**
+ * A namespace for various algorithms related to minimization over line.
+ */
+namespace Line_minimization
+{
+  /**
+   * Given $x\_low$ and $x\_hi$ together with values of functions
+   * ($f(x\_low)<f(x\_hi)$) and the gradient $g(x\_low)$, return the local
+   * minimum of the quadratic interpolation function.
+   */
+  template <typename NumberType>
+  boost::optional<NumberType>
+  quadratic_fit(const NumberType x_low,
```

Comment:
i will add a note on optional, but `NumberType` is here to be more general and support `float` or `long double`.

Reply:
I think the question I had was "I know it's possible to be more general, but is it useful?" Do you anticipate where either (i) double precision is not accurate enough, necessitating `long double`; or (ii) `double` is too slow because the function is called over and over, necessitating the use of `float`?

Reply:
> Do you anticipate where either (i) double precision is not accurate enough, necessitating long double; 

I think `long double` might be useful make the algorithm more stable.

Reply:
Hm. I have to say that I find that unlikely. In essence, you are solving a problem of finding a zero of a function (the derivative of your objective function) on the interval [0,1]. So the interval is neither small nor large, and the distance between the points where you are interpolating is likely also not. Now, if the function is extremely flat, then you may get into trouble, but that's going to point to a separate problem namely that your problem is ill-conditioned.

Either way, let's not argue and just leave it as is. Maybe someone would want to use AD types at one point or other.